doc/dev/adr/0023-property-support.md

```diff
@@ -0,0 +1,312 @@
+# ADR 23: SDK Support for Property
+
+## Context
+
+The AIO SDKs currently support two communication patterns: RPC and Telemetry.
+There is no support for the stateful communication pattern we call Property.
+There is an upcoming need to support this pattern because it is used heavily in OPC UA companion specifications, which AIO intends to support during [Br].
+
+Although one design option is to support Property without any SDK modifications, by providing support entirely within the ProtocolCompiler, this would be unsatisfactory for non-codegen scenarios.
+Another design option is for the SDKs to provide native support for Property that is parallel to &mdash; and independent of &mdash; their support for RPC and Telemetry; however, this would be very expensive in terms of development cost.
+
+## Decision
+
+The AIO SDKs will be enhanced with new envoy classes that support the Property communication pattern.
+Because the semantics of Property are decomposable into semantics of RPC and Telemetry, the new Property envoys will leverage the extant RPC and Telemetry envoys by layering functionality atop them.
+
+The ProtocolCompiler will be enhanced to support the DTDL Property content type.
+Generated code will target the new Property envoy classes in the SDKs.
+This requires an additive change to the DTDL Mqtt extension, and since version 4 of this extension has not officially shipped, this change will be rolled into the pending Mqtt extension version 4.
+
+## Property roles and semantics
+
+Just as RPC encompasses the two roles of Executor and Invoker, and as Telemetry encompasses the two roles of Sender and Receiver, the Property pattern encompasses two roles: Maintainer and Consumer.
+
+The Maintainer's responsibilities are to hold the Property state, to read and update the state upon request, and to configurably distribute notifications of state changes.
+
+The Consumer issues requests to the Maintainer, and it processes change notifications it receives from the Maintainer.
+
+The values of all Properties within an Interface must be readable atomically, writable atomically, and notifiable atomically.
+User code is not under any obligation to respect this atomicity, but the envoys are obligated to provide an API that readily supports atomicity if the application demands it.
+
+There are five specific actions that are relevant to Properties:
+
+* Action *write*:
+  * The Consumer sends a `write` request to the Maintainer, specifying new values for one or more Properties.
```

Comment:
The verbiage here strikes me as a bit odd. "Consumers" don't generally write. Should we refer to them as "Users" instead? Or "Actors"?

Reply:
It's a minor thing and I don't feel too strongly about it all. I was hesitant to even leave this comment since I can't think of a definitively better word for this "consumer" concept